tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py

```diff
@@ -131,6 +138,194 @@ def default_partition_fn(keys, shard_num):
+           math_ops.equal(de_now_shard_num, de_prev_shard_num), _do_nothing,
+           _traverse_files))
+  return traverse_files_result, keys_tensor_array_list, values_tensor_array_list
+
+
+def load_shard_from_file_system_or_tensor_array(
+    de_variable, shard, de_variable_folder_dir, traverse_files_result,
+    shard_idx, keys_tensor_array_list, values_tensor_array_list):
+  """Load DE keys/values files from file system or tensor array 
+      which generated from load_de_variable_from_file_system_to_tensor_array_if_shard_scale function. 
+    Load files directly when traverse_files_result == 0 or de_now_shard_num == 1.
+    Otherwith load tensor from keys_tensor_array_list and values_tensor_array_list.
+
+    de_variable: A present DynamicEmbedding Variable obeject.
+    shard: A present DynamicEmbedding Variable table obeject.
+    de_variable_folder_path: A directory path which saved DynamicEmbedding table shards before.
+    traverse_files_result: Generated from load_de_variable_from_file_system_to_tensor_array_if_shard_scale function.
+    shard_idx: Index of param shard in param de_variable.
+    keys_tensor_array_list: A list with keys partitions TF TensorArray. 
+      List length is present table shards number. TensorArray size is equal to files number.
+      Generated from load_de_variable_from_file_system_to_tensor_array_if_shard_scale function.
+    values_tensor_array_list: A list with values partitions TF TensorArray. 
+      List length is present table shards number. TensorArray size is equal to files number.
+      Generated from load_de_variable_from_file_system_to_tensor_array_if_shard_scale function.
+  Returns:
+    load_op: A TF Operation contains a flow to load shard table.
+  """
+
+  def load_de_table():
+    with ops.name_scope(de_variable.name, "%s_table_restore" % shard._name):
+      _DynamicEmbeddingShardSaveable_restore_op = shard.load_from_file_system(
+          de_variable_folder_dir, file_name=shard._name)
+    return _DynamicEmbeddingShardSaveable_restore_op
+
+  def upsert_de_table(upsert_keys_tensor_array_list,
+                      upsert_values_tensor_array_list):
+
+    def load_all_de_table():
+      with ops.name_scope(de_variable.name, "%s_table_restore" % shard._name):
+        return shard.load_from_file_system(de_variable_folder_dir,
+                                           file_name=shard._name,
+                                           load_entire_dir=True)
+
+    def insert_de_table(insert_keys_tensor_array_list,
+                        insert_values_tensor_array_list):
+      with ops.name_scope(de_variable.name, "%s_table_restore" % shard._name):
+        all_keys_partitions_tensor_array = insert_keys_tensor_array_list[
+            shard_idx]
+        all_values_partitions_tensor_array = insert_values_tensor_array_list[
+            shard_idx]
+
+        def sub_saveable_insert(i):
+          insert_op = shard.insert(all_keys_partitions_tensor_array.read(i),
```

Comment:
If this insert operation will cause the OOM on some workers? Do we have evaluated the risk here?

Reply:
> If this insert operation will cause the OOM on some workers? Do we have evaluated the risk here?

Yes, of course a big risk, and I don't have a good solution.
Besides this, there is no read_byte OP in TF, so I have to read the whole KV file which also may cause being OOM Killed.

Reply:
Already fix this by using TF FixedLengthRecordDataset to read buffer.